src/language.cpp

```diff
@@ -254,53 +248,50 @@ void Language::Clear()
  * @param base Base address of the strings.
  * @return Base offset for the registered strings. Add the index value of the \a names table to get the real string number.
  */
-uint16 Language::RegisterStrings(const TextData &td, const char * const names[], uint16 base)
-{
-	/* Count the number of strings. */
-	uint num_strings = 0;
-	const char * const *str = names;
-	while (*str != nullptr) {
-		num_strings++;
-		str++;
-	}
-
+uint16 LanguageManager::RegisterStrings(const TextData &td, const char * const names[], uint16 base)
+{
 	if (base == STR_GENERIC_END) {
-		/* Check for enough free space, and assign a block. */
-		if (this->first_free + num_strings >= STR_END_FREE_SPACE) {
+		base = this->free_index;
+		this->free_index += td.string_count;
+		if (this->free_index >= STR_END_FREE_SPACE) {
 			fprintf(stderr, "Not enough space to store strings.\n");
 			exit(1);
 		}
-		base = this->first_free;
-		this->first_free += num_strings;
 	} else {
-		/* Pre-defined strings should be completely below the free space. */
-		assert(base + num_strings >= base && base + num_strings <= this->first_free);
+		assert(base + td.string_count < this->free_index);
 	}
 
-	/* Copy strings in the expected order. */
-	uint16 number = base;
-	str = names;
-	while (*str != nullptr) {
-		this->registered[number] = nullptr;
-		for (uint i = 0; i < td.string_count; i++) {
-			const TextString *ts = td.strings.get() + i;
-			if (strcmp(ts->name, *str) == 0) {
-				this->registered[number] = ts;
+	const size_t old_size = this->string_names.size();
+	if (base + td.string_count > old_size) {
+		this->string_names.resize(base + td.string_count);
+		for (LanguageBundle &l : this->languages) {
+			assert(l.values.size() == old_size);
+			l.values.resize(base + td.string_count);
+		}
+	}
+
+	/* This would be much easier if names and text strings were in the same order. */
+	for (uint i = 0; i < td.string_count; ++i) {
+		assert(names[i] != nullptr && *names[i] != '\0');
+		this->string_names.at(base + i) = names[i];
+		for (uint j = 0; j < td.string_count; ++j) {
+			if (strcmp(names[i], td.strings[j].name) == 0) {
+				for (int l = 0; l < LANGUAGE_COUNT; ++l) this->languages[l].values.at(base + i) = td.strings[j].languages[l];
```

Comment:
Wrt the comment, you could build a `std::map<>` from `name` to `j` separately (before line 274), and use that to grab the `j` associated with an `i`.

A little longer, but you gain simplicity and speed here.